### PR TITLE
Dev zhi long path

### PR DESCRIPTION
**Pull request by zhili1208**
https://github.com/NuGet/Home/issues/6937
enabled long path support for nuget.exe
enabled long path support on CI machine.
@rrelyea 
@anangaur please review error message

### CODE REVIEW

File: scripts/cibuild/ConfigureVstsBuild.ps1 [hunk 1]
```diff
@@ -122,6 +124,12 @@ if (-not (Test-Path $regKeyNuGet) -or ($has32bitNode -and -not (Test-Path $regKe
     }
 }
 
+if (-not (Test-Path $regKeyFileSystem))
```
Comment (nkolev92) [inline scripts/cibuild/ConfigureVstsBuild.ps1:127]: should the build clean up at the end? 
This property will stay set on that machine.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx [hunk 2]
```diff
@@ -6158,4 +6158,13 @@ Oluşturma sırasında NuGet'in paketleri indirmesini önlemek için, Visual Stu
   <data name="ConsolePasswordProvider_PromptForPassword" xml:space="preserve">
     <value>Password: </value>
   </data>
+  <data name="Warning_LongPath_DisabledPolicy" xml:space="preserve">
+    <value>Please enbale long path support in local group policy, more details at https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/.</value>
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Typo: 
Please enable long path support in the local group policy, more details at https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/.

Should we use an aka.ms link because it'll be shorter? /cc @anangaur

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx [hunk 3]
```diff
@@ -6158,4 +6158,13 @@ Oluşturma sırasında NuGet'in paketleri indirmesini önlemek için, Visual Stu
...
+  </data>
+  <data name="Warning_LongPath_UnsupportedNetFramework" xml:space="preserve">
+    <value>Please install .net Framework 4.6.2 or later to support long path, more details at https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/.</value>
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Typo: 
Please install .net Framework 4.6.2 or later to support long paths, more details at https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/.

Same as above, I think we should make it an aka.ms link.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx [hunk 4]
```diff
@@ -6158,4 +6158,13 @@ Oluşturma sırasında NuGet'in paketleri indirmesini önlemek için, Visual Stu
...
+  </data>
+  <data name="Warning_LongPath_UnsupportOS" xml:space="preserve">
+    <value>Long Path is only supported on windows 10, more details at https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/.</value>
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Long paths are only supported on Windows 10....

Warning_LongPath_UnsupportOS => Warning_LongPath_UnsupportedOS

File: src/NuGet.Core/NuGet.PackageManagement/Utility/ExceptionUtility.cs [hunk 5]
```diff
@@ -12,20 +12,12 @@ public static Exception Unwrap(Exception exception)
...
+                throw new ArgumentNullException(nameof(exception));
             }
 
-            if (exception.InnerException == null)
+            while (exception.InnerException != null)
             {
-                return exception;
-            }
-
-            // Always return the inner exception from a target invocation exception
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.PackageManagement/Utility/ExceptionUtility.cs]: Why did you remove this?

**Review by nkolev92 — APPROVED** [review]
Minor comments on the language + 1 question regarding the unwrap exception code. 

LGTM 👏

File: src/NuGet.Core/NuGet.PackageManagement/Utility/ExceptionUtility.cs (hunk 5 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.PackageManagement/Utility/ExceptionUtility.cs]: this is incorrect for my change, It always return the root cause exception from any exception, not just invocation exception, also not the inner exception

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx (hunk 2 above)
Comment (anangaur) [reply src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Thats right. Here is the URL: https://aka.ms/nuget-long-path

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx (hunk 4 above)
Comment (anangaur) [reply src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Long file path is currently only supported on Windows 10. Fore more details, please refer to https://aka.ms/nuget-long-path

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx (hunk 3 above)
Comment (anangaur) [reply src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Please install .NET Framework 4.6.2 or above that supports long file paths. Fore more details, please refer to https://aka.ms/nuget-long-path

**Review by anangaur — APPROVED** [review]

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs [hunk 6]
```diff
@@ -342,5 +361,54 @@ private static void SetConsoleOutputEncoding(System.Text.Encoding encoding)
...
+            }
+            else if (!IsRuntimeGreaterThanNet462(logger))
+            {
+                logger.WriteWarning(LocalizedResourceManager.GetString(nameof(NuGetResources.Warning_LongPath_UnsupportedNetFramework)));
+            }
+        }
+
+        private static bool IsWindows10(ILogger logger)
+        {
+            var reg = RegistryKeyUtility.OpenSubKey(OSVersionRegistryKey, Registry.LocalMachine, logger);
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/Program.cs]: I'd expect it to open a regKey, read a specific value and close it right away. The current implementation leaves a loophole of keeping these regKeys open forever if there were intermittent exceptions.

**Review by jainaashish — CHANGES_REQUESTED** [review]
Change the way you read regKey and then close it, otherwise overall looks good.

File: scripts/cibuild/ConfigureVstsBuild.ps1 (hunk 1 above)
Comment (zhili1208) [reply scripts/cibuild/ConfigureVstsBuild.ps1:127]: I think it's ok to stay set there, it will be turned on by default in RS5

Comment (nkolev92) [reply scripts/cibuild/ConfigureVstsBuild.ps1:127]: Ahhh...true...and all the generic build pool machines should be RS4 soon enough.

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs [hunk 7]
```diff
@@ -342,5 +361,42 @@ private static void SetConsoleOutputEncoding(System.Text.Encoding encoding)
...
+            var longPathsEnabled = RegistryKeyUtility.GetValueFromRegistryKey("LongPathsEnabled", FilesystemRegistryKey, Registry.LocalMachine, logger);
+
+            return longPathsEnabled != null && (int)longPathsEnabled > 0;
+        }
+
+        private static bool IsRuntimeGreaterThanNet462(ILogger logger)
+        {
+            var release = RegistryKeyUtility.GetValueFromRegistryKey("Release", DotNetSetupRegistryKey, Registry.LocalMachine, logger);
+
+            return release != null && (int)release >= 394802;
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/Program.cs]: what is this magic number 394802? can you create a constant for that?

File: src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+            try
+            {
+                var key = registryKey?.OpenSubKey(name);
+
+                var result = key?.GetValue(name);
+
+                if (key != null)
+                {
+                    // Note that according to MSDN, this method does nothing if you call it on an instance of RegistryKey that is already closed.
+                    key.Close();
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs]: shouldn't this `key.Close()` be replicated in catch block as well? so that if key is not closed when exception is thrown, it can be closed in catch block.

**Review by jainaashish — APPROVED** [review]
Looks good.
